streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +39,678 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
 
+/**
+ * ChangelogReader is created and maintained by the stream thread and used for both updating standby tasks and
+ * restoring active tasks. It manages the restore consumer, including its assigned partitions, when to pause / resume
+ * these partitions, etc.
+ *
+ * The reader also maintains the source of truth for restoration state: only active tasks restoring changelog could
+ * be completed, while standby tasks updating changelog would always be in restoring state after being initialized.
+ */
 public class StoreChangelogReader implements ChangelogReader {
 
+    enum ChangelogState {
+        // registered but need to be initialized (i.e. set its starting, end, limit offsets)
+        REGISTERED("REGISTERED"),
+
+        // initialized and restoring
+        RESTORING("RESTORING"),
+
+        // completed restoring (only for active restoring task, standby task should never be completed)
+        COMPLETED("COMPLETED");
+
+        public final String name;
+
+        ChangelogState(final String name) {
+            this.name = name;
+        }
+    }
+
+    // NOTE we assume that the changelog read is used only for either 1) restoring active task
+    // or 2) updating standby task at a given time, but never doing both
+    enum ChangelogReaderState {
+        ACTIVE_RESTORING("ACTIVE_RESTORING"),
+
+        STANDBY_UPDATING("STANDBY_UPDATING");
+
+        public final String name;
+
+        ChangelogReaderState(final String name) {
+            this.name = name;
+        }
+    }
+
+    private class ChangelogMetadata {
+
+        private final TopicPartition changelogPartition;
+
+        private final ProcessorStateManager stateManager;
+
+        private ChangelogState changelogState;
+
+        private long totalRestored;
+
+        // only for active restoring tasks (for standby changelog it is null)
+        // NOTE we do not book keep the current offset since we leverage state manager as its source of truth
+        private Long restoreEndOffset;
+
+        // only for standby tasks that use source topics as changelogs (for active it is null);
+        // if it is not on source topics it is also null
+        private Long restoreLimitOffset;
+
+        // buffer records polled by the restore consumer;
+        private List<ConsumerRecord<byte[], byte[]>> bufferedRecords;
+
+        // the limit index (exclusive) inside the buffered records beyond which should not be used to restore
+        // either due to limit offset (standby) or committed end offset (active)
+        private int bufferedLimitIndex;
+
+        private ChangelogMetadata(final TopicPartition changelogPartition, final ProcessorStateManager stateManager) {
+            this.stateManager = stateManager;
```

Comment:
This is a meta reply: I agree this is a cleaner way with less code to register the store-metadata instead of state manager and perhaps consolidate store-metadata with changelog-metadata too, but on motivation for doing the clean up is to have clear separation of source-of-truth ownership between classes, and I worry that consolidating them would make one class (ChangelogReader) to be too large and a bit clumsy for reading / debugging purposes.

I agree that the interaction between state manager and changelog reader is pretty tight (as you see, there are some functions that are exclusively used by changelog reader), but I feel the current class boundary cut makes it a bit easier to reason. WDYT?

Reply:
I see your intent. What about this:
* Make `StateStoreMetadata` read-only for `StoreChangelogReader` by declare `setOffset()` as `private`.
* Pass objects of `StateStoreMetadata` to `StoreChangelogReader` at registration time.
* Remove `changelogPartition` as field in `ChangelogMetadata`.
* Keep `stateManager` as field in `ChangelogMetadata`.
* Change `void restore(final TopicPartition changelogPartition, final List<ConsumerRecord<byte[], byte[]>> restoreRecords)` to `void restore(final StateStoreMetadata storeMetadata, final List<ConsumerRecord<byte[], byte[]>> restoreRecords)` (forget my comment about moving `ProcessorStateManager#restore()` to `StoreChangelogReader`).

In this way only `ProcessorStateManager` is allowed to change the offsets, i.e., it is responsible for `StateStoreMetadata` and stays the source of truth. `StoreChangelogReader` does not need to make lookups into maps in `ProcessorStateManager` because it can read the `StateStoreMetadata` for each changelog but is not allowed to write it. Furthermore, we do not need `mustFindNode()` because `StoreChangelogReader` naturally passes the correct `StateStoreMetadata` to `ProcessorStateManager#restore()`.

I have also a renaming proposal for `StateStoreMetadata`. Since it contains also a `StateStore`, metadata is not really an appropriate name. I would call it `ManagedStateStore` since each object of this class represent a state store that is managed by the `ProcessorStateManager`.

Reply:
Thanks Bruno, I'll try it out in the next PR.